### PR TITLE
[Feature] HVU Training

### PR DESCRIPTION
**Pull request by kennymckormick**
Support HVU Training

### CODE REVIEW

File: mmaction/datasets/hvu_dataset.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+from mmcv.utils import print_log
+
+from ..core import mean_average_precision
+from .base import BaseDataset
+from .registry import DATASETS
+
+
+@DATASETS.register_module()
+class HVUDataset(BaseDataset):
+    """HVU dataset, which support the recognition tags of multiple categories.
```
Comment (dreamerlin) [inline mmaction/datasets/hvu_dataset.py:14]: supports

File: mmaction/datasets/hvu_dataset.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    Accept both video annotation files or rawframe annotation files.
+
+    The dataset loads videos or raw frames and apply specified transforms to
```
Comment (dreamerlin) [inline mmaction/datasets/hvu_dataset.py:17]: applies

File: mmaction/datasets/hvu_dataset.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        }
+
+
+    Args:
+        ann_file (str): Path to the annotation file, should be a json file.
+        pipeline (list[dict | callable]): A sequence of data transforms.
+        tag_categories (list[str]): List of category names of tags.
+        tag_category_nums (list[int]): List of number of tags in each category.
+        **kwargs: Keyword arguments for ``BaseDataset``.
+    """
```
Comment (dreamerlin) [inline mmaction/datasets/hvu_dataset.py:60]: missing args description for `filename_tmpl`.

File: mmaction/datasets/hvu_dataset.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+        assert len(tag_categories) == len(tag_category_nums)
+        self.tag_categories = tag_categories
+        self.tag_category_nums = tag_category_nums
+        self.filename_tmpl = filename_tmpl
+        self.num_categories = len(self.tag_categories)
+        self.num_tags = sum(self.tag_category_nums)
+        self.category2num = {
+            k: v
+            for k, v in zip(tag_categories, tag_category_nums)
+        }
```
Comment (dreamerlin) [inline mmaction/datasets/hvu_dataset.py:78]: `self.category2num = dict(zip(tag_categories, tag_category_nums))`

File: mmaction/datasets/hvu_dataset.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+        self.start_idx = [0]
+        for i in range(self.num_categories - 1):
+            self.start_idx.append(self.start_idx[-1] +
+                                  self.tag_category_nums[i])
+        self.category2startidx = {
+            k: v
+            for k, v in zip(tag_categories, self.start_idx)
+        }
```
Comment (dreamerlin) [inline mmaction/datasets/hvu_dataset.py:86]: `self.category2startidx = dict(zip(tag_categories, self.start_idx))`

File: mmaction/datasets/hvu_dataset.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+        self.start_index = 0
+        if 'start_index' in kwargs:
+            self.start_index = kwargs.pop('start_index')
```
Comment (dreamerlin) [inline mmaction/datasets/hvu_dataset.py:89]: `self.start_index = kwargs.pop('start_index', 0)`

File: mmaction/datasets/pipelines/loading.py [hunk 7]
```diff
@@ -6,13 +6,71 @@
...
+    def __init__(self, **kwargs):
+        self.hvu_initialized = False
+
+    def init_hvu_info(self, categories, category_nums):
+        assert len(categories) == len(category_nums)
+        self.categories = categories
+        self.category_nums = category_nums
+        self.num_categories = len(self.categories)
+        self.num_tags = sum(self.category_nums)
+        self.category2num = {k: v for k, v in zip(categories, category_nums)}
```
Comment (dreamerlin) [inline mmaction/datasets/pipelines/loading.py:34]: dict(zip)

File: mmaction/datasets/pipelines/loading.py [hunk 8]
```diff
@@ -6,13 +6,71 @@
...
+        self.start_idx = [0]
+        for i in range(self.num_categories - 1):
+            self.start_idx.append(self.start_idx[-1] + self.category_nums[i])
+        self.category2startidx = {
+            k: v
+            for k, v in zip(categories, self.start_idx)
+        }
```
Comment (dreamerlin) [inline mmaction/datasets/pipelines/loading.py:41]: dict(zip)

File: mmaction/models/losses/hvu_loss.py [hunk 9]
```diff
@@ -0,0 +1,126 @@
+import torch  # isort: skip
+import torch.nn.functional as F  # isort: skip
+
+from ..registry import LOSSES  # isort: skip
+from .base import BaseWeightedLoss  # isort: skip
```
Comment (dreamerlin) [inline mmaction/models/losses/hvu_loss.py:5]: why `isort :skip` here ?

File: mmaction/models/losses/hvu_loss.py [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+class HVULoss(BaseWeightedLoss):
+    """Calculate the BCELoss for HVU.
+
+    Args:
+        categories (list[str]): Names of tag categories, tags are organized in
+            this order. Default: ['action', 'attribute', 'concept', 'event',
+            'object', 'scene'].
+        category_nums (list[int]): Number of tags for each category. Default:
+            [739, 117, 291, 69, 1679, 248].
+        category_loss_weights (list[float]): Loss weights of categories, it
```
Comment (dreamerlin) [inline mmaction/models/losses/hvu_loss.py:18]: `List`, as a mutable data type, may cause danger when set as input args. you can use `tuple`

File: mmaction/models/losses/hvu_loss.py [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+            applies only if `loss_type == 'individual'`. The loss weights will
+            be normalized so that the sum equals to 1, so that you can give any
+            positive number as loss weight. Default: [1, 1, 1, 1, 1, 1].
+        loss_type (str): The loss type we calculate, we can either calculate
+            the BCELoss for all tags, or calculate the BCELoss for tags in each
+            category. Choices are ['individual', 'all']. Default: 'all'.
```
Comment (dreamerlin) [inline mmaction/models/losses/hvu_loss.py:24]: Choices are 'individual' or 'all'

File: mmaction/datasets/hvu_dataset.py [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+            start_idx = self.category2startidx[category]
+            num = self.category2num[category]
+            preds = [
+                result[start_idx:start_idx + num]
+                for video_idx, result in enumerate(results)
+                if category in gt_labels[video_idx]
+            ]
+            gts = [
+                gt_label[category]
+                for video_idx, gt_label in enumerate(gt_labels)
```
Comment (dreamerlin) [inline mmaction/datasets/hvu_dataset.py:163]: `for gt_label in gt_labels`

File: mmaction/datasets/hvu_dataset.py [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+                if category in gt_label
+            ]
+
+            # convert label list to ndarray
+            def label2array(label):
```
Comment (dreamerlin) [inline mmaction/datasets/hvu_dataset.py:168]: I think this function should be written as a `staticmethod`

File: mmaction/models/losses/hvu_loss.py [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+        Returns:
+            torch.Tensor: The returned CrossEntropy loss.
+        """
+        assert 'mask' in aux_info
+        assert 'category_mask' in aux_info
+        mask = aux_info['mask']
+        category_mask = aux_info['category_mask']
+
+        if self.loss_type == 'all':
+            loss_cls = F.binary_cross_entropy_with_logits(
```
Comment (dreamerlin) [inline mmaction/models/losses/hvu_loss.py:77]: https://github.com/open-mmlab/mmaction2/blob/master/mmaction/models/losses/cross_entropy_loss.py#L28

File: mmaction/models/losses/hvu_loss.py [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+                return dict(loss_cls=torch.mean(loss_cls))
+        elif self.loss_type == 'individual':
+            losses = {}
+            loss_weights = {}
+            for name, num, start_idx in zip(self.categories,
+                                            self.category_nums,
+                                            self.category_startidx):
+                category_score = cls_score[:, start_idx:start_idx + num]
+                category_label = label[:, start_idx:start_idx + num]
+                category_loss = F.binary_cross_entropy_with_logits(
```
Comment (dreamerlin) [inline mmaction/models/losses/hvu_loss.py:96]: https://github.com/open-mmlab/mmaction2/blob/master/mmaction/models/losses/cross_entropy_loss.py#L28

File: mmaction/models/recognizers/recognizer2d.py [hunk 16]
```diff
@@ -6,7 +6,7 @@
 class Recognizer2D(BaseRecognizer):
     """2D recognizer model framework."""
 
-    def forward_train(self, imgs, labels):
+    def forward_train(self, imgs, labels, aux_info):
```
Comment (dreamerlin) [inline mmaction/models/recognizers/recognizer2d.py:9]: can this `aux_info` thing be replaced with `**kwargs`? since there might be more additional keywargs for users to add.

File: mmaction/models/losses/hvu_loss.py (hunk 14 above)
Comment (kennymckormick) [reply mmaction/models/losses/hvu_loss.py:77]: I think it's OK to use F.binary_cross_entropy_with_logits, switching to BCELossWithLogits doesn't make the code more concise. Anyway, BCELossWithLogits is just another wrapper of F.binary_cross_entropy_with_logits

File: mmaction/models/losses/hvu_loss.py (hunk 15 above)
Comment (kennymckormick) [reply mmaction/models/losses/hvu_loss.py:96]: Same as the last comment

**Comment by dreamerlin** [comment]
please resolve the conflict to trigger the unittest check

**Comment by dreamerlin** [comment]
`HVULoss` and `LoadHVULabel` should be unittested.

**Comment by dreamerlin** [comment]
See https://codecov.io/gh/open-mmlab/mmaction2/pull/235/diff?src=pr&el=tree#diff-bW1hY3Rpb24vZGF0YXNldHMvYmFzZS5weQ== for detail

**Comment by innerlee** [comment]
ok, will look at it tomorrow

File: mmaction/core/evaluation/accuracy.py [hunk 17]
```diff
@@ -104,17 +104,46 @@ def top_k_accuracy(scores, labels, topk=(1, )):
     return res
 
 
+def mmit_mean_average_precision(scores, labels):
```
Comment (innerlee) [inline mmaction/core/evaluation/accuracy.py:107]: does this metric have a more `neutral` name?

Comment (kennymckormick) [reply mmaction/core/evaluation/accuracy.py:107]: Might be hard, since I only find that the MMIT challenge uses this definition for mean average precision. All the others use the normal definition like detection.

**Review by innerlee — APPROVED** [review]

**Comment by innerlee** [comment]
linting error
